File: question-generator-service/simulation/simulate_students.go
// simulate_students.go
// Large-Scale Student Simulation Framework - Go Implementation
// Phase 2.2 Load Testing with configurable concurrency

package main

import (
	"bytes"
	"context"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"os"
	"sync"
	"time"
)

// Configuration for simulation
type SimulationConfig struct {
	APIBaseURL      string
	StudentCount    int
	ConcurrentUsers int
	TestDuration    time.Duration
	RequestsPerUser int
	OutputFile      string
	Verbose         bool
}

// Student simulation profile
type VirtualStudent struct {
	ID          string
	ExamType    string
	Subject     string
	Format      string
	Difficulty  float64
	Performance float64 // Simulated ability level
}

// Request/Response structures
type QuestionRequest struct {
	StudentID           string  `json:"student_id"`
	TopicID            string  `json:"topic_id"`
	ExamType           string  `json:"exam_type"`
	Subject            string  `json:"subject"`
	Format             string  `json:"format"`
	RequestedDifficulty float64 `json:"requested_difficulty"`
	SessionID          string  `json:"session_id"`
	RequestID          string  `json:"request_id"`
}

type QuestionResponse struct {
	QuestionID     string                 `json:"question_id"`
	QuestionText   string                 `json:"question_text"`
	Options        map[string]string      `json:"options"`
	CorrectAnswer  string                 `json:"correct_answer"`
	Difficulty     float64               `json:"difficulty"`
	GenerationTime int64                 `json:"generation_time"`
	QualityScore   float64               `json:"quality_score"`
	Status         string                `json:"status"`
	Metadata       map[string]interface{} `json:"metadata"`
}

// Performance metrics
type StudentMetrics struct {
	StudentID          string
	TotalRequests      int
	SuccessfulRequests int
	FailedRequests     int
	AvgResponseTime    time.Duration
	MinResponseTime    time.Duration
	MaxResponseTime    time.Duration
	TotalResponseTime  time.Duration
	ErrorMessages      []string
	QuestionsAnswered  int
	CorrectAnswers     int
	Accuracy           float64
}

// Global metrics collector
type MetricsCollector struct {
	mutex           sync.RWMutex
	StudentMetrics  map[string]*StudentMetrics
	StartTime       time.Time
	EndTime         time.Time
	TotalRequests   int64
	SuccessRequests int64
	ErrorRequests   int64
	ResponseTimes   []time.Duration
}

func NewMetricsCollector() *MetricsCollector {
	return &MetricsCollector{
		StudentMetrics: make(map[string]*StudentMetrics),
		StartTime:      time.Now(),
		ResponseTimes:  make([]time.Duration, 0),
	}
}

func (mc *MetricsCollector) RecordRequest(studentID string, responseTime time.Duration, success bool, errorMsg string) {
	mc.mutex.Lock()
	defer mc.mutex.Unlock()

	mc.TotalRequests++
	mc.ResponseTimes = append(mc.ResponseTimes, responseTime)

	if success {
		mc.SuccessRequests++
	} else {
		mc.ErrorRequests++
	}

	// Update student-specific metrics
	if _, exists := mc.StudentMetrics[studentID]; !exists {
		mc.StudentMetrics[studentID] = &StudentMetrics{
			StudentID:       studentID,
			MinResponseTime: responseTime,
			MaxResponseTime: responseTime,
			ErrorMessages:   make([]string, 0),
		}
	}

	student := mc.StudentMetrics[studentID]
	student.TotalRequests++
	student.TotalResponseTime += responseTime

	if success {
		student.SuccessfulRequests++
	} else {
		student.FailedRequests++
		if errorMsg != "" {
			student.ErrorMessages = append(student.ErrorMessages, errorMsg)
		}
	}

	if responseTime < student.MinResponseTime {
		student.MinResponseTime = responseTime
	}
	if responseTime > student.MaxResponseTime {
		student.MaxResponseTime = responseTime
	}

	student.AvgResponseTime = student.TotalResponseTime / time.Duration(student.TotalRequests)
}

func (mc *MetricsCollector) RecordAnswer(studentID string, correct bool) {
	mc.mutex.Lock()
	defer mc.mutex.Unlock()

	if student, exists := mc.StudentMetrics[studentID]; exists {
		student.QuestionsAnswered++
		if correct {
			student.CorrectAnswers++
		}
		student.Accuracy = float64(student.CorrectAnswers) / float64(student.QuestionsAnswered)
	}
}

func (mc *MetricsCollector) GetSummary() map[string]interface{} {
	mc.mutex.RLock()
	defer mc.mutex.RUnlock()

	mc.EndTime = time.Now()
	duration := mc.EndTime.Sub(mc.StartTime)

	// Calculate percentiles
	var p50, p95, p99 time.Duration
	if len(mc.ResponseTimes) > 0 {
		// Simple percentile calculation (for production use proper sorting)
		total := len(mc.ResponseTimes)
		p50 = mc.ResponseTimes[total*50/100]
		p95 = mc.ResponseTimes[total*95/100]
		p99 = mc.ResponseTimes[total*99/100]
	}

	var avgResponseTime time.Duration
	if len(mc.ResponseTimes) > 0 {
		var total time.Duration
		for _, rt := range mc.ResponseTimes {
			total += rt
		}
		avgResponseTime = total / time.Duration(len(mc.ResponseTimes))
	}

	return map[string]interface{}{
		"simulation_duration":    duration.Seconds(),
		"total_requests":        mc.TotalRequests,
		"successful_requests":   mc.SuccessRequests,
		"failed_requests":       mc.ErrorRequests,
		"success_rate":          float64(mc.SuccessRequests) / float64(mc.TotalRequests) * 100,
		"requests_per_second":   float64(mc.TotalRequests) / duration.Seconds(),
		"avg_response_time_ms":  avgResponseTime.Milliseconds(),
		"p50_response_time_ms":  p50.Milliseconds(),
		"p95_response_time_ms":  p95.Milliseconds(),
		"p99_response_time_ms":  p99.Milliseconds(),
		"concurrent_users":      len(mc.StudentMetrics),
		"start_time":           mc.StartTime.Format(time.RFC3339),
		"end_time":             mc.EndTime.Format(time.RFC3339),
	}
}

// Generate realistic virtual student profiles
func GenerateVirtualStudents(count int) []VirtualStudent {
	rand.Seed(time.Now().UnixNano())
	
	examTypes := []string{"JEE_MAIN", "JEE_ADVANCED", "NEET", "FOUNDATION"}
	subjects := []string{"PHYSICS", "CHEMISTRY", "MATHEMATICS", "BIOLOGY"}
	formats := []string{"MCQ", "NUMERICAL", "ASSERTION_REASON", "PASSAGE"}
	
	students := make([]VirtualStudent, count)
	
	for i := 0; i < count; i++ {
		examType := examTypes[rand.Intn(len(examTypes))]
		
		// Subject selection based on exam type
		var availableSubjects []string
		if examType == "NEET" {
			availableSubjects = []string{"PHYSICS", "CHEMISTRY", "BIOLOGY"}
		} else if examType == "JEE_MAIN" || examType == "JEE_ADVANCED" {
			availableSubjects = []string{"PHYSICS", "CHEMISTRY", "MATHEMATICS"}
		} else {
			availableSubjects = subjects
		}
		
		subject := availableSubjects[rand.Intn(len(availableSubjects))]
		format := formats[rand.Intn(len(formats))]
		
		// Normal distribution for difficulty preference and performance
		difficulty := 0.3 + rand.Float64()*0.4 // Between 0.3 and 0.7
		performance := 0.2 + rand.Float64()*0.6 // Between 0.2 and 0.8
		
		students[i] = VirtualStudent{
			ID:          fmt.Sprintf("sim_student_%05d", i+1),
			ExamType:    examType,
			Subject:     subject,
			Format:      format,
			Difficulty:  difficulty,
			Performance: performance,
		}
	}
	
	return students
}

// Simulate a single student's behavior
func SimulateStudent(ctx context.Context, config SimulationConfig, student VirtualStudent, 
	metrics *MetricsCollector, wg *sync.WaitGroup) {
	defer wg.Done()

	client := &http.Client{
		Timeout: 30 * time.Second,
	}

	for i := 0; i < config.RequestsPerUser; i++ {
		select {
		case <-ctx.Done():
			return
		default:
			// Generate realistic question request
			request := QuestionRequest{
				StudentID:           student.ID,
				TopicID:            generateTopicID(student.Subject),
				ExamType:           student.ExamType,
				Subject:            student.Subject,
				Format:             student.Format,
				RequestedDifficulty: student.Difficulty + (rand.Float64()-0.5)*0.2, // ±0.1 variation
				SessionID:          fmt.Sprintf("session_%s_%d", student.ID, i),
				RequestID:          fmt.Sprintf("req_%s_%d_%d", student.ID, i, time.Now().UnixNano()),
			}

			// Measure response time
			startTime := time.Now()
			success, errorMsg := makeQuestionRequest(client, config.APIBaseURL, request)
			responseTime := time.Since(startTime)

			// Record metrics
			metrics.RecordRequest(student.ID, responseTime, success, errorMsg)

			if config.Verbose {
				status := "SUCCESS"
				if !success {
					status = "FAILED: " + errorMsg
				}
				log.Printf("Student %s: Request %d/%d - %s (%s)", 
					student.ID, i+1, config.RequestsPerUser, status, responseTime)
			}

			// Simulate answer submission based on student performance
			if success && rand.Float64() < 0.8 { // 80% answer questions
				correct := rand.Float64() < student.Performance
				metrics.RecordAnswer(student.ID, correct)
			}

			// Realistic interval between requests (1-5 seconds)
			time.Sleep(time.Duration(1000+rand.Intn(4000)) * time.Millisecond)
		}
	}
}

// Make HTTP request to question generation API
func makeQuestionRequest(client *http.Client, baseURL string, request QuestionRequest) (bool, string) {
	requestBody, err := json.Marshal(request)
	if err != nil {
		return false, fmt.Sprintf("JSON marshal error: %v", err)
	}

	url := fmt.Sprintf("%s/v1/questions/generate", baseURL)
	
	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(requestBody))
	if err != nil {
		return false, fmt.Sprintf("Request creation error: %v", err)
	}

	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Accept", "application/json")
	httpReq.Header.Set("X-Request-ID", request.RequestID)

	resp, err := client.Do(httpReq)
	if err != nil {
		return false, fmt.Sprintf("HTTP request error: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return false, fmt.Sprintf("HTTP %d: %s", resp.StatusCode, resp.Status)
	}

	// Parse response to validate
	var response QuestionResponse
	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
		return false, fmt.Sprintf("Response parsing error: %v", err)
	}

	return true, ""
}

// Generate realistic topic IDs based on subject
func generateTopicID(subject string) string {
	topics := map[string][]string{
		"PHYSICS": {
			"PHY_MECHANICS_KINEMATICS", "PHY_MECHANICS_DYNAMICS", "PHY_THERMODYNAMICS",
			"PHY_ELECTROMAGNETISM", "PHY_OPTICS", "PHY_MODERN_PHYSICS",
		},
		"CHEMISTRY": {
			"CHEM_PHYSICAL_CHEMISTRY", "CHEM_ORGANIC_CHEMISTRY", "CHEM_INORGANIC_CHEMISTRY",
			"CHEM_CHEMICAL_BONDING", "CHEM_THERMODYNAMICS", "CHEM_ATOMIC_STRUCTURE",
		},
		"MATHEMATICS": {
			"MATH_ALGEBRA", "MATH_CALCULUS", "MATH_TRIGONOMETRY",
			"MATH_COORDINATE_GEOMETRY", "MATH_PROBABILITY", "MATH_VECTORS",
		},
		"BIOLOGY": {
			"BIO_CELL_BIOLOGY", "BIO_GENETICS", "BIO_ECOLOGY",
			"BIO_HUMAN_PHYSIOLOGY", "BIO_PLANT_PHYSIOLOGY", "BIO_EVOLUTION",
		},
	}

	subjectTopics := topics[subject]
	return subjectTopics[rand.Intn(len(subjectTopics))]
}

// Export metrics to CSV
func ExportMetricsToCSV(metrics *MetricsCollector, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("create file error: %w", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write header
	header := []string{
		"student_id", "total_requests", "successful_requests", "failed_requests",
		"success_rate", "avg_response_time_ms", "min_response_time_ms", "max_response_time_ms",
		"questions_answered", "correct_answers", "accuracy", "error_count",
	}
	writer.Write(header)

	// Write student data
	metrics.mutex.RLock()
	defer metrics.mutex.RUnlock()

	for _, student := range metrics.StudentMetrics {
		successRate := float64(student.SuccessfulRequests) / float64(student.TotalRequests) * 100
		
		record := []string{
			student.StudentID,
			fmt.Sprintf("%d", student.TotalRequests),
			fmt.Sprintf("%d", student.SuccessfulRequests),
			fmt.Sprintf("%d", student.FailedRequests),
			fmt.Sprintf("%.2f", successRate),
			fmt.Sprintf("%d", student.AvgResponseTime.Milliseconds()),
			fmt.Sprintf("%d", student.MinResponseTime.Milliseconds()),
			fmt.Sprintf("%d", student.MaxResponseTime.Milliseconds()),
			fmt.Sprintf("%d", student.QuestionsAnswered),
			fmt.Sprintf("%d", student.CorrectAnswers),
			fmt.Sprintf("%.2f", student.Accuracy*100),
			fmt.Sprintf("%d", len(student.ErrorMessages)),
		}
		writer.Write(record)
	}

	return nil
}

func main() {
	// Command-line flags
	var config SimulationConfig
	flag.StringVar(&config.APIBaseURL, "url", "http://localhost:8080", "API base URL")
	flag.IntVar(&config.StudentCount, "students", 100, "Number of virtual students")
	flag.IntVar(&config.ConcurrentUsers, "concurrent", 10, "Concurrent users")
	flag.DurationVar(&config.TestDuration, "duration", 5*time.Minute, "Test duration")
	flag.IntVar(&config.RequestsPerUser, "requests", 10, "Requests per user")
	flag.StringVar(&config.OutputFile, "output", "simulation_results.csv", "Output CSV file")
	flag.BoolVar(&config.Verbose, "verbose", false, "Verbose logging")
	flag.Parse()

	log.Printf("Starting Student Simulation with %d students, %d concurrent users", 
		config.StudentCount, config.ConcurrentUsers)
	
	// Generate virtual students
	students := GenerateVirtualStudents(config.StudentCount)
	
	// Initialize metrics collector
	metrics := NewMetricsCollector()
	
	// Create context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), config.TestDuration)
	defer cancel()
	
	// Start simulation
	var wg sync.WaitGroup
	semaphore := make(chan struct{}, config.ConcurrentUsers)
	
	for _, student := range students {
		wg.Add(1)
		semaphore <- struct{}{} // Acquire semaphore
		
		go func(s VirtualStudent) {
			defer func() { <-semaphore }() // Release semaphore
			SimulateStudent(ctx, config, s, metrics, &wg)
		}(student)
	}
	
	// Wait for completion or timeout
	wg.Wait()
	
	// Generate summary
	summary := metrics.GetSummary()
	log.Println("\n=== SIMULATION SUMMARY ===")
	for key, value := range summary {
		log.Printf("%s: %v", key, value)
	}
	
	// Export detailed metrics
	if err := ExportMetricsToCSV(metrics, config.OutputFile); err != nil {
		log.Printf("Error exporting metrics: %v", err)
	} else {
		log.Printf("Detailed metrics exported to: %s", config.OutputFile)
	}
	
	log.Println("Simulation completed successfully!")
}

File: question-generator-service/simulation/simulate_students.py
#!/usr/bin/env python3
"""
simulate_students.py
Large-Scale Student Simulation Framework - Python Implementation
Phase 2.2 Load Testing with asyncio and configurable concurrency
"""

import asyncio
import aiohttp
import argparse
import csv
import json
import logging
import random
import time
import uuid
from dataclasses import dataclass, field
from typing import List, Dict, Optional, Tuple
from datetime import datetime, timedelta
import statistics

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

@dataclass
class SimulationConfig:
    """Configuration for simulation parameters."""
    api_base_url: str = "http://localhost:8080"
    student_count: int = 1000
    concurrent_users: int = 50
    test_duration: int = 300  # seconds
    requests_per_user: int = 20
    output_file: str = "simulation_results_python.csv"
    verbose: bool = False
    request_timeout: int = 30

@dataclass
class VirtualStudent:
    """Represents a virtual student with realistic characteristics."""
    id: str
    exam_type: str
    subject: str
    format: str
    difficulty: float
    performance: float  # Simulated ability level (0.0-1.0)
    learning_rate: float = field(default_factory=lambda: random.uniform(0.1, 0.3))
    stress_tolerance: float = field(default_factory=lambda: random.uniform(0.4, 0.9))

@dataclass
class RequestMetrics:
    """Metrics for a single request."""
    student_id: str
    request_id: str
    success: bool
    response_time: float
    error_message: Optional[str] = None
    timestamp: datetime = field(default_factory=datetime.now)

@dataclass
class StudentSummary:
    """Summary metrics for a single student."""
    student_id: str
    total_requests: int = 0
    successful_requests: int = 0
    failed_requests: int = 0
    total_response_time: float = 0.0
    min_response_time: float = float('inf')
    max_response_time: float = 0.0
    questions_answered: int = 0
    correct_answers: int = 0
    errors: List[str] = field(default_factory=list)
    
    @property
    def success_rate(self) -> float:
        if self.total_requests == 0:
            return 0.0
        return (self.successful_requests / self.total_requests) * 100
    
    @property
    def avg_response_time(self) -> float:
        if self.total_requests == 0:
            return 0.0
        return self.total_response_time / self.total_requests
    
    @property
    def accuracy(self) -> float:
        if self.questions_answered == 0:
            return 0.0
        return (self.correct_answers / self.questions_answered) * 100

class MetricsCollector:
    """Collects and analyzes simulation metrics."""
    
    def __init__(self):
        self.start_time = time.time()
        self.end_time = None
        self.request_metrics: List[RequestMetrics] = []
        self.student_summaries: Dict[str, StudentSummary] = {}
        self._lock = asyncio.Lock()
    
    async def record_request(self, metrics: RequestMetrics):
        """Record metrics for a single request."""
        async with self._lock:
            self.request_metrics.append(metrics)
            
            # Update student summary
            if metrics.student_id not in self.student_summaries:
                self.student_summaries[metrics.student_id] = StudentSummary(
                    student_id=metrics.student_id
                )
            
            summary = self.student_summaries[metrics.student_id]
            summary.total_requests += 1
            summary.total_response_time += metrics.response_time
            
            if metrics.success:
                summary.successful_requests += 1
            else:
                summary.failed_requests += 1
                if metrics.error_message:
                    summary.errors.append(metrics.error_message)
            
            # Update response time stats
            if metrics.response_time < summary.min_response_time:
                summary.min_response_time = metrics.response_time
            if metrics.response_time > summary.max_response_time:
                summary.max_response_time = metrics.response_time
    
    async def record_answer(self, student_id: str, correct: bool):
        """Record answer submission."""
        async with self._lock:
            if student_id in self.student_summaries:
                summary = self.student_summaries[student_id]
                summary.questions_answered += 1
                if correct:
                    summary.correct_answers += 1
    
    def get_summary(self) -> Dict:
        """Generate comprehensive simulation summary."""
        self.end_time = time.time()
        duration = self.end_time - self.start_time
        
        # Calculate aggregate metrics
        total_requests = len(self.request_metrics)
        successful_requests = sum(1 for m in self.request_metrics if m.success)
        failed_requests = total_requests - successful_requests
        
        response_times = [m.response_time for m in self.request_metrics]
        
        if response_times:
            avg_response_time = statistics.mean(response_times)
            p50_response_time = statistics.median(response_times)
            p95_response_time = statistics.quantiles(response_times, n=20)[18]  # 95th percentile
            p99_response_time = statistics.quantiles(response_times, n=100)[98]  # 99th percentile
        else:
            avg_response_time = p50_response_time = p95_response_time = p99_response_time = 0
        
        return {
            'simulation_duration': duration,
            'total_requests': total_requests,
            'successful_requests': successful_requests,
            'failed_requests': failed_requests,
            'success_rate': (successful_requests / total_requests * 100) if total_requests > 0 else 0,
            'requests_per_second': total_requests / duration if duration > 0 else 0,
            'avg_response_time_ms': avg_response_time * 1000,
            'p50_response_time_ms': p50_response_time * 1000,
            'p95_response_time_ms': p95_response_time * 1000,
            'p99_response_time_ms': p99_response_time * 1000,
            'concurrent_users': len(self.student_summaries),
            'start_time': datetime.fromtimestamp(self.start_time).isoformat(),
            'end_time': datetime.fromtimestamp(self.end_time).isoformat(),
        }

def generate_virtual_students(count: int) -> List[VirtualStudent]:
    """Generate realistic virtual student profiles."""
    exam_types = ["JEE_MAIN", "JEE_ADVANCED", "NEET", "FOUNDATION"]
    formats = ["MCQ", "NUMERICAL", "ASSERTION_REASON", "PASSAGE"]
    
    students = []
    
    for i in range(count):
        exam_type = random.choice(exam_types)
        
        # Subject selection based on exam type
        if exam_type == "NEET":
            subjects = ["PHYSICS", "CHEMISTRY", "BIOLOGY"]
        elif exam_type in ["JEE_MAIN", "JEE_ADVANCED"]:
            subjects = ["PHYSICS", "CHEMISTRY", "MATHEMATICS"]
        else:
            subjects = ["PHYSICS", "CHEMISTRY", "MATHEMATICS", "BIOLOGY"]
        
        subject = random.choice(subjects)
        format_type = random.choice(formats)
        
        # Realistic distributions
        difficulty = random.uniform(0.2, 0.8)
        performance = random.normalvariate(0.5, 0.15)  # Normal distribution
        performance = max(0.1, min(0.9, performance))  # Clamp to [0.1, 0.9]
        
        student = VirtualStudent(
            id=f"py_student_{i+1:05d}",
            exam_type=exam_type,
            subject=subject,
            format=format_type,
            difficulty=difficulty,
            performance=performance
        )
        
        students.append(student)
    
    return students

def generate_topic_id(subject: str) -> str:
    """Generate realistic topic ID based on subject."""
    topics = {
        "PHYSICS": [
            "PHY_MECHANICS_KINEMATICS", "PHY_MECHANICS_DYNAMICS", "PHY_THERMODYNAMICS",
            "PHY_ELECTROMAGNETISM", "PHY_OPTICS", "PHY_MODERN_PHYSICS", "PHY_WAVES",
        ],
        "CHEMISTRY": [
            "CHEM_PHYSICAL_CHEMISTRY", "CHEM_ORGANIC_CHEMISTRY", "CHEM_INORGANIC_CHEMISTRY",
            "CHEM_CHEMICAL_BONDING", "CHEM_THERMODYNAMICS", "CHEM_ATOMIC_STRUCTURE",
        ],
        "MATHEMATICS": [
            "MATH_ALGEBRA", "MATH_CALCULUS", "MATH_TRIGONOMETRY",
            "MATH_COORDINATE_GEOMETRY", "MATH_PROBABILITY", "MATH_VECTORS",
        ],
        "BIOLOGY": [
            "BIO_CELL_BIOLOGY", "BIO_GENETICS", "BIO_ECOLOGY",
            "BIO_HUMAN_PHYSIOLOGY", "BIO_PLANT_PHYSIOLOGY", "BIO_EVOLUTION",
        ],
    }
    
    return random.choice(topics[subject])

async def make_question_request(
    session: aiohttp.ClientSession, 
    api_url: str, 
    request_data: Dict
) -> Tuple[bool, Optional[str], Dict]:
    """Make HTTP request to question generation API."""
    try:
        async with session.post(
            f"{api_url}/v1/questions/generate",
            json=request_data,
            headers={
                "Content-Type": "application/json",
                "Accept": "application/json",
                "X-Request-ID": request_data["request_id"]
            }
        ) as response:
            if response.status == 200:
                response_data = await response.json()
                return True, None, response_data
            else:
                error_text = await response.text()
                return False, f"HTTP {response.status}: {error_text}", {}
    
    except asyncio.TimeoutError:
        return False, "Request timeout", {}
    except Exception as e:
        return False, f"Request error: {str(e)}", {}

async def simulate_student(
    student: VirtualStudent,
    config: SimulationConfig,
    metrics: MetricsCollector,
    semaphore: asyncio.Semaphore,
    session: aiohttp.ClientSession
):
    """Simulate a single student's behavior."""
    async with semaphore:  # Limit concurrency
        for request_num in range(config.requests_per_user):
            # Generate request
            request_data = {
                "student_id": student.id,
                "topic_id": generate_topic_id(student.subject),
                "exam_type": student.exam_type,
                "subject": student.subject,
                "format": student.format,
                "requested_difficulty": student.difficulty + random.uniform(-0.1, 0.1),
                "session_id": f"session_{student.id}_{request_num}",
                "request_id": f"req_{student.id}_{request_num}_{int(time.time()*1000)}"
            }
            
            # Measure response time
            start_time = time.time()
            success, error_msg, response_data = await make_question_request(
                session, config.api_base_url, request_data
            )
            response_time = time.time() - start_time
            
            # Record metrics
            request_metrics = RequestMetrics(
                student_id=student.id,
                request_id=request_data["request_id"],
                success=success,
                response_time=response_time,
                error_message=error_msg
            )
            
            await metrics.record_request(request_metrics)
            
            if config.verbose:
                status = "SUCCESS" if success else f"FAILED: {error_msg}"
                logger.info(
                    f"Student {student.id}: Request {request_num+1}/{config.requests_per_user} - "
                    f"{status} ({response_time:.3f}s)"
                )
            
            # Simulate answer submission
            if success and random.random() < 0.8:  # 80% answer questions
                correct = random.random() < student.performance
                await metrics.record_answer(student.id, correct)
            
            # Realistic delay between requests
            await asyncio.sleep(random.uniform(1, 5))

async def run_simulation(config: SimulationConfig) -> MetricsCollector:
    """Run the complete simulation."""
    logger.info(f"Starting simulation with {config.student_count} students, "
                f"{config.concurrent_users} concurrent users")
    
    # Generate virtual students
    students = generate_virtual_students(config.student_count)
    
    # Initialize metrics collector
    metrics = MetricsCollector()
    
    # Create semaphore to limit concurrency
    semaphore = asyncio.Semaphore(config.concurrent_users)
    
    # Create HTTP session with timeout
    timeout = aiohttp.ClientTimeout(total=config.request_timeout)
    
    async with aiohttp.ClientSession(timeout=timeout) as session:
        # Create tasks for all students
        tasks = [
            simulate_student(student, config, metrics, semaphore, session)
            for student in students
        ]
        
        # Run simulation with timeout
        try:
            await asyncio.wait_for(
                asyncio.gather(*tasks, return_exceptions=True),
                timeout=config.test_duration
            )
        except asyncio.TimeoutError:
            logger.warning("Simulation timed out - collecting partial results")
    
    return metrics

def export_metrics_to_csv(metrics: MetricsCollector, filename: str):
    """Export detailed metrics to CSV file."""
    with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
        fieldnames = [
            'student_id', 'total_requests', 'successful_requests', 'failed_requests',
            'success_rate', 'avg_response_time_ms', 'min_response_time_ms', 'max_response_time_ms',
            'questions_answered', 'correct_answers', 'accuracy', 'error_count'
        ]
        
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
        writer.writeheader()
        
        for summary in metrics.student_summaries.values():
            writer.writerow({
                'student_id': summary.student_id,
                'total_requests': summary.total_requests,
                'successful_requests': summary.successful_requests,
                'failed_requests': summary.failed_requests,
                'success_rate': f"{summary.success_rate:.2f}",
                'avg_response_time_ms': f"{summary.avg_response_time * 1000:.2f}",
                'min_response_time_ms': f"{summary.min_response_time * 1000:.2f}",
                'max_response_time_ms': f"{summary.max_response_time * 1000:.2f}",
                'questions_answered': summary.questions_answered,
                'correct_answers': summary.correct_answers,
                'accuracy': f"{summary.accuracy:.2f}",
                'error_count': len(summary.errors)
            })

async def main():
    """Main function with argument parsing."""
    parser = argparse.ArgumentParser(description="Student Load Testing Simulation - Python")
    parser.add_argument('--url', default='http://localhost:8080', help='API base URL')
    parser.add_argument('--students', type=int, default=1000, help='Number of virtual students')
    parser.add_argument('--concurrent', type=int, default=50, help='Concurrent users')
    parser.add_argument('--duration', type=int, default=300, help='Test duration in seconds')
    parser.add_argument('--requests', type=int, default=20, help='Requests per user')
    parser.add_argument('--output', default='simulation_results_python.csv', help='Output CSV file')
    parser.add_argument('--verbose', action='store_true', help='Verbose logging')
    parser.add_argument('--timeout', type=int, default=30, help='Request timeout in seconds')
    
    args = parser.parse_args()
    
    # Configure logging level
    if args.verbose:
        logging.getLogger().setLevel(logging.DEBUG)
    
    # Create configuration
    config = SimulationConfig(
        api_base_url=args.url,
        student_count=args.students,
        concurrent_users=args.concurrent,
        test_duration=args.duration,
        requests_per_user=args.requests,
        output_file=args.output,
        verbose=args.verbose,
        request_timeout=args.timeout
    )
    
    # Run simulation
    try:
        metrics = await run_simulation(config)
        
        # Generate and display summary
        summary = metrics.get_summary()
        logger.info("\n=== SIMULATION SUMMARY ===")
        for key, value in summary.items():
            logger.info(f"{key}: {value}")
        
        # Export detailed metrics
        export_metrics_to_csv(metrics, config.output_file)
        logger.info(f"Detailed metrics exported to: {config.output_file}")
        
        logger.info("Simulation completed successfully!")
        
    except Exception as e:
        logger.error(f"Simulation failed: {e}")
        raise

if __name__ == "__main__":
    # Run the async main function
    asyncio.run(main())